### ISSUE TITLE
Assembly (project) conventions should support Directory.Build.Props & Directory.Build.Targets

### ISSUE THREAD

**Issue by flakey-bit** [andrewabest/Conventional#88]
MSBuild allows templating / "meta-programming" of `csproj` files via `Directory.Build.Props` & `Target.Build.Props` - see this page for more [background](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022) but basically 

* `Directory.Build.Props` allows setting default property values across all projects in the directory (recursively)
* `Target.Build.Props` allows setting _dependant_ property values (i.e. including customizations from individual projects) 

At least that's my understanding - haven't used `Target.Build.Props` myself.

Existing conventions like `MustNotReferenceDllsFromTransientOrSdkDirectoriesConventionSpecification` that parse the project XML are unaware of this functionality and are arguably broken as a result.

For example, a `Directory.Build.Props` file could add an (assembly) Reference w/ a HintPath to **all** projects in the directory, which `MustNotReferenceDllsFromTransientOrSdkDirectoriesConventionSpecification` would be unaware of when examing the project XML, leading to false-negative.

**Comment by flakey-bit** [comment]
Ironically, `Directory.Build.Props` & `Directory.Build.Targets` files make **keeping things consistent** _easier_ - however they make **checking things are consistent** _harder_ 🤔

**Comment by andrewabest** [comment]
Good spotting @flakey-bit. Fixing this would be reasonably straight-forward. I'll either get to it eventually, or someone can pick it up, I've labelled it a good first issue if someone happens to come by and wants to contribute.

**Comment by flakey-bit** [comment]
> Fixing this would be reasonably straight-forward

@andrewabest do you have some thoughts in mind as to how we could address this?

**Comment by andrewabest** [comment]
I was thinking you would essentially have to replicate what MSBuild does - for any project, locate the directory.build.props file closest to it, and consider its values in applying the convention ([docs](https://learn.microsoft.com/en-us/visualstudio/msbuild/customize-by-directory?view=vs-2022))

It can get a bit complex, but we could support the simpler version of it: don't support multiple props files, just support the nearest one per project.

**Comment by flakey-bit** [comment]
I _think_ you're suggesting that we should fully evaluate the project XML and pass the "composed" `csproj` (i.e. post application of `Directory.Build.Props` & `Directory.Build.Targets`)  to Conventional conventions (in other words, treating it as a templating system).

If so, I agree 👍

It would be neat if we could do so by _leveraging_ MSBuild, rather than reimplementing it. If we can find a way to do that I guess we'd get full support (multiple props files etc) "for free".

**Comment by andrewabest** [comment]
Yup, strong agree there. I know dotnet-affected uses MsBuild internally ([example](https://github.com/leonardochaia/dotnet-affected/blob/main/src/DotnetAffected.Core/Processor/AffectedProcessor.cs)), so it is very possible that that is a doable thing.
